### PR TITLE
msetup: Allow (re)configure of not empty builddir

### LINKED ISSUE #11418: Build directory is allowed to be parent directory of project
**Describe the bug**
It is possible, for any project containing meson.build, to configure the parent directory as the build directory. This is easy to do by accident (if you create a build subdirectory and forget to change into it). Critically, 'meson setup --wipe' may then erase the project itself, and any other sibling project.

**To Reproduce**
`meson setup ..`

**Expected behavior**
This should be rejected with an error message. The build directory should not be allowed to be the parent directory of the project containing the `meson.build` file.

Moreover, there seems to be general confusion about the semantics of the directory argument to `meson setup <dir>` since it seems the following are accepted:

`<dir>` is the build directory, current working directory is the project (i.e. contains meson.build)
`<dir>` is the project directory, current working directory is the build directory

While I can understand the occasional convenience of allowing the 2nd form, it doesn't seem to be documented and the fact that it works leads to confusion. The whole bug would be avoided if the 2nd form was rejected, as the documentation implies it should be.

What makes the bug even worse is that `meson setup --wipe ..` will happily erase your project directory, and anything else that happens to be in the same parent directory. It's pretty easy to accidentally configure the parent as build dir i.e.:

1. `meson setup ..` -  whoops, forgot to change into the build subdir first, parent is now configured as build dir. Let's suppose I don't realise, but do see that I wanted a different configuration. Or even that I do realise, but forget to remove the various stuff that meson created in the parent directory so it remains configured as a build directory, then some time later make a similar mistake of forgetting to change into build directory before I do:
2. `meson setup .. -wipe` - whoops, the whole project is gone and so is any other project that happened to be in the parent directory.

I just got hit by this bug. I'm pretty sure I haven't lost anything important, and I realise I'm somewhat to blame, but the fact that this can happen so easily is not good.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?
Native build, but probably it doesn't matter.
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)
Linux.
* what Python version are you using e.g. 3.8.0
~~3.7.5~~ 3.11.1
* what `meson --version`
1.0.0
* what `ninja --version` if it's a Ninja build
N/A, not necessary to run ninja to reproduce.

### PR DESCRIPTION
**Pull request by xclaesse**
This fix regression caused by https://github.com/mesonbuild/meson/pull/11431 and also integrate changes from https://github.com/mesonbuild/meson/pull/11590 and https://github.com/mesonbuild/meson/pull/11427.

Fixes #11418
Closes #11427
Closes #11590

### CODE REVIEW

**Review by bruchar1 — APPROVED** [review]

**Comment by bruchar1** [comment]
Fixes #11418

**Comment by lithomas1** [comment]
Thanks for the quick fix!
I'll try and pull this locally later this week to check that this works.

**Comment by xclaesse** [comment]
Ok, the CI failure is really impressive:
- We were using `raise SystemExit` with no parameter in the "Directory already configured" case, which actually make meson return success.
- That had the side effect that the 2nd run of the loop in test_preprocessor_checks_CPPFLAGS (when testing with CFLAGS) succeeds because it hits the "Directory already configured" case which is considered success.
- This PR changed SystemExit to MesonException which caused that test to now fail as it should. Thus adding `self.new_builddir()`.
- But now that test fails because cc.get_define() never has read CFLAGS, only CPPFLAGS.

File: mesonbuild/msetup.py [hunk 1]
```diff
@@ -169,16 +169,15 @@ def validate_dirs(self, dir1: str, dir2: str, reconfigure: bool, wipe: bool) ->
         has_partial_build = os.path.isdir(priv_dir)
         if has_valid_build:
             if not reconfigure and not wipe:
-                print('Directory already configured.\n'
-                      '\nJust run your build command (e.g. ninja) and Meson will regenerate as necessary.\n'
+                raise MesonException('Directory already configured.\n\n'
+                      'Just run your build command (e.g. ninja) and Meson will regenerate as necessary.\n'
```
Comment (eli-schwartz) [inline mesonbuild/msetup.py:176]: This eventuality MUST NOT be considered user error.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: mesonbuild/msetup.py (hunk 1 above)
Comment (xclaesse) [reply mesonbuild/msetup.py:176]: Why not? That makes no sense.

Comment (xclaesse) [reply mesonbuild/msetup.py:176]: Hmm, looks like 5years ago that was a problem already: https://github.com/mesonbuild/meson/pull/4249. That still feels completely wrong IMHO.

Comment (xclaesse) [reply mesonbuild/msetup.py]: Ok, kept that as success for now because it's really unrelated to this PR anyway. But IMHO we should reconsider this in the future.

File: mesonbuild/msetup.py [hunk 2]
```diff
@@ -169,16 +172,18 @@ def validate_dirs(self, dir1: str, dir2: str, reconfigure: bool, wipe: bool) ->
         has_partial_build = os.path.isdir(priv_dir)
         if has_valid_build:
             if not reconfigure and not wipe:
-                print('Directory already configured.\n'
-                      '\nJust run your build command (e.g. ninja) and Meson will regenerate as necessary.\n'
+                print('Directory already configured.\n\n'
```
Comment (bruchar1) [inline mesonbuild/msetup.py:175]: `mlog.warning`?

Comment (xclaesse) [reply mesonbuild/msetup.py:175]: I think it deserves to be a warning, and probably to be hard error in the future, because it ignores all -Dfoo options. OTOH, now I'm wondering why having --reconfigure flag at all, that could just be the default behaviour.

Comment (xclaesse) [reply mesonbuild/msetup.py]: Added a commit to make --reconfigure the default, that flag is thus useless now.

**Comment by heitbaum** [comment]
👍 Tested 1.1.0rc1 along with this PR and [LibreELEC/LibreELEC.tv](https://github.com/LibreELEC/LibreELEC.tv) builds. Versus `Directory is not empty and does not contain a previous build:` without the PR

**Comment by xclaesse** [comment]
> Added a commit to make --reconfigure the default, that flag is thus useless now.

Dropped that commit because we are in RC and it's not the time to reinvent our CLI. I'll make a separate PR for that.

**Comment by heitbaum** [comment]
> > Added a commit to make --reconfigure the default, that flag is thus useless now.
> 
> Dropped that commit because we are in RC and it's not the time to reinvent our CLI. I'll make a separate PR for that.

Retested. Working and fixes the regression in 1.1.0rc1

File: mesonbuild/msetup.py [hunk 3]
```diff
@@ -169,16 +169,18 @@ def validate_dirs(self, dir1: str, dir2: str, reconfigure: bool, wipe: bool) ->
...
-                      'to force Meson to regenerate.\n'
-                      '\nIf build failures persist, run "meson setup --wipe" to rebuild from scratch\n'
-                      'using the same options as passed when configuring the build.'
-                      '\nTo change option values, run "meson configure" instead.')
+                      'to force Meson to regenerate.\n\n'
+                      'If build failures persist, run "meson setup --wipe" to rebuild from scratch\n'
+                      'using the same options as passed when configuring the build.\n'
+                      'To change option values, run "meson configure" instead.')
+                # This is not an error for backward compatibility:
+                # https://github.com/mesonbuild/meson/pull/4249
```
Comment (eli-schwartz) [inline mesonbuild/msetup.py:183]: IMHO the comment is wrong -- this is not backwards compat, just sensible behavior.

Independent of what we do when rerunning "setup" and having it do nothing, you do have another PR to make it do something if `--options` are passed, which is another perhaps reasonable thing to do here... but at the core of it I still argue that this is "intended UX", not "preserved for backwards compat".

**Review by eli-schwartz — COMMENTED** [review]
The commit message for the first commit seems outdated.

File: mesonbuild/msetup.py (hunk 3 above)
Comment (xclaesse) [reply mesonbuild/msetup.py:183]: It should be hard error or set new options, that can be argued either way. But current behaviour of ignoring options is totally wrong. I wrote that comment when I thought we should be heading toward hard error, but I changed my mind since then. I'll put more nuance into that comment, indeed.

Comment (eli-schwartz) [reply mesonbuild/msetup.py]: The nuance is... better... although as I said, I still disagree with your "should either", but whatever. :P  

However, the update is in the wrong commit, currently you say one thing in the first commit and then change your mind in the unrelated second commit!

Also, the commit message for the second commit (still) should be fixed to state what it's actually doing now.

Comment (xclaesse) [reply mesonbuild/msetup.py]: done
